### PR TITLE
MRG: Lower mem usage rfft

### PR DESCRIPTION
**Pull request by larsoner**
This PR, by using `np.fft.rfft` for filtering and resampling:

- Reduces memory slightly
- Should speed up filtering a little bit
- Fixes some minor bugs with `resample(..., window=<non-boxcar>)`

Only reason it's not MRG is that I am not on a CUDA machine currently so can't see if that passes (though it should). @jona-sassenhagen if you are on a CUDA machine can you run `pytest mne/tests/test_filter.py` and let me know if it passes for you?

### CODE REVIEW

**Comment by jona-sassenhagen** [comment]
Hm ... it seems I broke my pycuda. Tensorflow still working, pycuda not so much. So far I haven't managed to repair it. Sorry.

**Comment by jona-sassenhagen** [comment]
Ok, I fixed it, I'll try ...

**Comment by jona-sassenhagen** [comment]
Here's my error log:

[fft.txt](https://github.com/mne-tools/mne-python/files/2280328/fft.txt)

There's a few problems with the dimensionality of returned arrays.

**Comment by larsoner** [comment]
Okay @jona-sassenhagen if you have a minute feel free to try again

**Comment by jona-sassenhagen** [comment]
Better :)

```
$pytest mne/tests/test_filter.py
================================================= test session starts =================================================
platform linux -- Python 3.6.6, pytest-3.7.1, py-1.5.4, pluggy-0.7.1
rootdir: /home/jona/tools/mne-python, inifile: setup.cfg
plugins: cov-2.5.1, hypothesis-3.59.1
collected 15 items

mne/tests/test_filter.py .s..........F..                                                                        [100%]

====================================================== FAILURES =======================================================
______________________________________________________ test_cuda ______________________________________________________

    def test_cuda():
        """Test CUDA-based filtering."""
        # NOTE: don't make test_cuda() the last test, or pycuda might spew
        # some warnings about clean-up failing
        # Also, using `n_jobs='cuda'` on a non-CUDA system should be fine,
        # as it should fall back to using n_jobs=1.
        sfreq = 500
        sig_len_secs = 20
        a = rng.randn(sig_len_secs * sfreq)
        kwargs = dict(fir_design='firwin')

        with catch_logging() as log_file:
            for fl in ['auto', '10s', 2048]:
                args = [a, sfreq, 4, 8, None, fl, 1.0, 1.0]
                bp = filter_data(*args, **kwargs)
                bp_c = filter_data(*args, n_jobs='cuda', verbose='info', **kwargs)
>               assert_array_almost_equal(bp, bp_c, 12)
E               AssertionError:
E               Arrays are not almost equal to 12 decimals
E
E               (mismatch 99.98%)
E                x: array([ 2.634611279140e-16,  6.886644789747e-02,  1.372661039115e-01, ...,
E                      -5.546386239411e-02, -2.787588055490e-02, -2.844946500602e-16])
E                y: array([ 0.000000000000e+00,  1.377160847494e-01,  2.744986996809e-01, ...,
E                      -1.109141854589e-01, -5.574495630557e-02, -3.053113317719e-16])

a          = array([-1.10444724,  1.04429969,  0.83388901, ...,  0.97168444,
        0.66623804,  0.82323362])
args       = [array([-1.10444724,  1.04429969,  0.83388901, ...,  0.97168444,
        0.66623804,  0.82323362]), 500, 4, 8, None, 'auto', ...]
bp         = array([ 2.63461128e-16,  6.88664479e-02,  1.37266104e-01, ...,
       -5.54638624e-02, -2.78758806e-02, -2.84494650e-16])
bp_c       = array([ 0.00000000e+00,  1.37716085e-01,  2.74498700e-01, ...,
       -1.10914185e-01, -5.57449563e-02, -3.05311332e-16])
fl         = 'auto'
kwargs     = {'fir_design': 'firwin'}
log_file   = <_io.StringIO object at 0x7fb31aa94318>
sfreq      = 500
sig_len_secs = 20

mne/tests/test_filter.py:502: AssertionError
=============================================== short test summary info ===============================================
SKIP [1] mne/tests/test_filter.py:32: Test test_mne_c_design skipped, requires MNE-C.
============================================== slowest 20 test durations ==============================================
2.83s call     mne/tests/test_filter.py::test_resample_scipy
1.52s call     mne/tests/test_filter.py::test_resamp_stim_channel
0.93s call     mne/tests/test_filter.py::test_filters
0.84s call     mne/tests/test_filter.py::test_1d_filter
0.11s call     mne/tests/test_filter.py::test_notch_filters
0.03s call     mne/tests/test_filter.py::test_filter_auto
0.01s call     mne/tests/test_filter.py::test_cuda
0.01s call     mne/tests/test_filter.py::test_resample
0.01s call     mne/tests/test_filter.py::test_iir_stability
0.01s call     mne/tests/test_filter.py::test_estimate_ringing
0.01s call     mne/tests/test_filter.py::test_filter_array
0.00s call     mne/tests/test_filter.py::test_resample_raw
0.00s call     mne/tests/test_filter.py::test_detrend
0.00s call     mne/tests/test_filter.py::test_interp2
0.00s teardown mne/tests/test_filter.py::test_resample_scipy
0.00s teardown mne/tests/test_filter.py::test_filters
0.00s setup    mne/tests/test_filter.py::test_filter_array
0.00s setup    mne/tests/test_filter.py::test_resamp_stim_channel
0.00s setup    mne/tests/test_filter.py::test_filter_auto
0.00s teardown mne/tests/test_filter.py::test_resamp_stim_channel
=================================== 1 failed, 13 passed, 1 skipped in 6.70 seconds ====================================
```

**Comment by jona-sassenhagen** [comment]
```
$pytest mne/tests/test_filter.py
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.6.6, pytest-3.7.1, py-1.5.4, pluggy-0.7.1
rootdir: /home/jona/tools/mne-python, inifile: setup.cfg
plugins: cov-2.5.1, hypothesis-3.59.1
collected 15 items

mne/tests/test_filter.py .s.............                                                                                                                                                                                               [100%]
========================================================================================================== short test summary info ===========================================================================================================
SKIP [1] mne/tests/test_filter.py:32: Test test_mne_c_design skipped, requires MNE-C.

========================================================================================================= slowest 20 test durations ==========================================================================================================
2.83s call     mne/tests/test_filter.py::test_resample_scipy
1.55s call     mne/tests/test_filter.py::test_resamp_stim_channel
0.94s call     mne/tests/test_filter.py::test_filters
0.84s call     mne/tests/test_filter.py::test_1d_filter
0.23s call     mne/tests/test_filter.py::test_cuda
0.11s call     mne/tests/test_filter.py::test_notch_filters
0.03s call     mne/tests/test_filter.py::test_filter_auto
0.01s call     mne/tests/test_filter.py::test_resample
0.01s call     mne/tests/test_filter.py::test_iir_stability
0.01s call     mne/tests/test_filter.py::test_estimate_ringing
0.01s call     mne/tests/test_filter.py::test_filter_array
0.00s call     mne/tests/test_filter.py::test_resample_raw
0.00s call     mne/tests/test_filter.py::test_detrend
0.00s call     mne/tests/test_filter.py::test_interp2
0.00s setup    mne/tests/test_filter.py::test_filter_array
0.00s teardown mne/tests/test_filter.py::test_resample_scipy
0.00s teardown mne/tests/test_filter.py::test_filters
0.00s setup    mne/tests/test_filter.py::test_resamp_stim_channel
0.00s teardown mne/tests/test_filter.py::test_notch_filters
0.00s teardown mne/tests/test_filter.py::test_resamp_stim_channel
==================================================================================================== 14 passed, 1 skipped in 6.94 seconds ====================================================================================================
```

**Comment by larsoner** [comment]
Okay this is ready for merge then. Once this is done I can move onto https://github.com/mne-tools/mne-python/pull/4119#issuecomment-412087576

File: mne/cuda.py [hunk 1]
```diff
@@ -106,19 +106,18 @@ def init_cuda(ignore_config=False):
 ###############################################################################
 # Repeated FFT multiplication
 
-def setup_cuda_fft_multiply_repeated(n_jobs, h_fft):
+def setup_cuda_fft_multiply_repeated(n_jobs, h, n_fft):
```
Comment (jona-sassenhagen) [inline mne/cuda.py]: shouldn't this function be private? (ie, the API is changed below ...)

**Comment by jona-sassenhagen** [comment]
I can't really review the code cause the math is too heavy for me ... but as you can see, it passes tests on my machine.

File: mne/cuda.py (hunk 1 above)
Comment (larsoner) [reply mne/cuda.py]: Yes maybe I should change them to private. They weren't ever really meant to be public and aren't in our API docs so hopefully it's not too bad to do it

**Comment by agramfort** [comment]
thx @larsoner !

**Pull request merged**